### PR TITLE
[PT2E][Quant] Use subgraph matcher annotate linear pattern

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99580
* #99399
* #99398

This diff adds subgraph matcher for pattern matching. Furthermore, we also move
annotations for the matched subgraph in a way that only input and output nodes
of the matched subgraph have quantization related valid annotations.

Differential Revision: [D45133867](https://our.internmc.facebook.com/intern/diff/D45133867/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @soumith @desertfire

### CODE REVIEW

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py [hunk 1]
```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
...
+            bias_node = None
+            for ph in weight_or_bias:
+                weight_or_bias = getattr(gm, ph.target)
+                if weight_or_bias.ndim == 2:
+                    weight_node = ph
+                if weight_or_bias.ndim == 1:
+                    bias_node = ph
+
+            # bias and output act
+            if _is_annotated([act_node]) == False:
```
Comment (jerryzh168) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:355]: nit: `not _is_annotated([act_node])`?

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py [hunk 2]
```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
...
+                act_node.meta["target_dtype_info"] = {
+                    "input_act_obs_or_fq_ctr": None,
+                    "output_act_obs_or_fq_ctr": get_act_obs_or_fq_ctr(
+                        quantization_config
+                    ),
+                    "weight_obs_or_fq_ctr": None,
+                    "bias_obs_or_fq_ctr": None,
```
Comment (jerryzh168) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:362]: these are optional I think, you can remove them

**Review by jerryzh168 — APPROVED** [review]
LGTM, should be able to cleanup the annotation further after https://github.com/pytorch/pytorch/pull/99220 is landed

File: torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py [hunk 3]
```diff
@@ -278,52 +311,88 @@ def _annotate_conv2d(
...
+                weight_or_bias = getattr(gm, ph.target)
+                if weight_or_bias.ndim == 2:
+                    weight_node = ph
+                if weight_or_bias.ndim == 1:
+                    bias_node = ph
+
+            # bias and output act
+            if _is_annotated([act_node]) == False:
+                act_node.meta["target_dtype_info"] = {
+                    "input_act_obs_or_fq_ctr": None,
```
Comment (leslie-fang-intel) [inline torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:357]: May I ask a question why `input_act_obs_or_fq_ctr` is None here for `act_node`? Does it mean we don't insert observer for the input of  `act_node`?

Comment (jerryzh168) [reply torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py:357]: `act_node` here is actually the node that produces the activation for linear, so the output of `act_node` is the input for linear

**Comment by kimishpatel** [comment]
> LGTM, should be able to cleanup the annotation further after #99220 is landed

Yes, that would be good, but lets catch up on what that means. I explicitly annotated in between nodes without anything,  because I expected quant workflow to not touch those nodes. But I want to sync with you whether thats a valid assumption.

**Comment by jerryzh168** [comment]
> > LGTM, should be able to cleanup the annotation further after #99220 is landed
> 
> Yes, that would be good, but lets catch up on what that means. I explicitly annotated in between nodes without anything, because I expected quant workflow to not touch those nodes. But I want to sync with you whether thats a valid assumption.

yeah you can annotate with None to mean don't touch this, after my PR you also don't need to specify anything and it will just work (assumed to be torch.float after that PR)

**Pull request closed without merging**